h/templates/deform/form.jinja2

```diff
@@ -5,39 +5,50 @@
 {% set form_buttons_class = 'form-actions-buttons' %}
 {% set form_message_class = 'form-actions-message' %}
 {% endif %}
-
+{% if field.use_inline_editing %}
+{% set form_controller_class = 'js-form' %}
+{% endif %}
 <form id="{{ field.formid }}"
       action="{{ field.action }}"
       method="{{ field.method }}"
       enctype="multipart/form-data"
       accept-charset="utf-8"
-      class="form {% if field.css_class %}{{ field.css_class }} {% endif %}">
+      class="form {{ field.css_class or '' }} {{ form_controller_class }}">
   <input type="hidden" name="__formid__" value="{{ field.formid }}" />
 
+  <div class="form__backdrop" data-ref="formBackdrop"></div>
+
   {%- for f in field.children -%}
     {{ field.renderer(field.widget.item_template, field=f, cstruct=cstruct.get(f.name, null)) }}
   {% endfor -%}
 
-  <div class="form-actions">
+  <div class="form-actions {% if field.use_inline_editing %} u-js-hide-when-loading {% endif %}"
+       data-ref="formActions">
+    <div class="form__submit-error" data-ref="formSubmitError">
+      <span>{% trans %}Unable to save changes: {% endtrans %}</span>
+      <span data-ref="formSubmitErrorMessage"></span>
+    </div>
     <div class="{{ form_message_class }}">
       {%- if field.footer %}{{ field.footer | safe }}{% endif -%}
     </div>
     {% if feature('activity_pages') %}
     <div class="u-stretch"></div>
     {% endif %}
-    <div class="{{ form_buttons_class }}">
+    <div class="{{ form_buttons_class }}" data-ref="formButtons">
```

Comment:
Ah sorry, this is left over from an earlier iteration. Same for `formBtn`. I'll remove both.